bodhi/tests/server/functional/test_updates.py

```diff
@@ -2520,6 +2520,99 @@ def test_edit_testing_update_reset_karma_with_same_tester(self, publish, *args):
+        up.comment(self.db, u'Failed to work', author=u'ralph', karma=-1)
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        self.assertTrue(up.has_negative_karma)
+
+        # The same user gives positive karma later
+        up.comment(self.db, u'wfm', author=u'ralph', karma=1)
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        self.assertFalse(up.has_negative_karma)
+        self.assertEquals(up.karma, 1)
+
+    @mock.patch(**mock_valid_requirements)
+    @mock.patch('bodhi.server.notifications.publish')
+    def test_has_no_negative_karma(self, publish, *args):
+        """
+        The test asserts False for has_negative_karma i.e,
+        if update has no negative karma
+        """
+        user = User(name=u'bob')
+        self.db.add(user)
+        self.db.flush()
+
+        nvr = 'bodhi-2.1-1.fc17'
+        args = self.get_update(nvr)
+        resp = self.app.post_json('/updates/', args).json_body
+        publish.assert_called_with(topic='update.request.testing', msg=ANY)
+
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        up.request = None
+        up.status = UpdateStatus.testing
+        self.db.flush()
+
+        up.comment(self.db, u'LGTM', author=u'mac', karma=1)
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        self.assertFalse(up.has_negative_karma)
+
+        # Karma with no comment
+        up.comment(self.db, u' ', author=u'bowlofeggs', karma=1)
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        self.assertFalse(up.has_negative_karma)
+        self.assertEquals(up.karma, 2)
+
+    @mock.patch(**mock_valid_requirements)
+    @mock.patch('bodhi.server.notifications.publish')
+    def test_disable_autopush_for_critical_updates(self, publish, *args):
+        """Make sure that autopush is disabled if critical update receives any negative karma"""
+        user = User(name=u'bob')
+        self.db.add(user)
+        self.db.flush()
+
+        nvr = u'kernel-3.11.5-300.fc17'
+        args = self.get_update(nvr)
+        args['autokarma'] = True
+        resp = self.app.post_json('/updates/', args)
+        self.assertTrue(resp.json['critpath'])
+        self.assertEquals(resp.json['request'], 'testing')
+        publish.assert_called_with(topic='update.request.testing', msg=ANY)
+
+        upd = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        upd.status = UpdateStatus.testing
+        upd.request = None
+        self.db.flush()
+
+        upd.comment(self.db, u'Failed to work', author=u'ralph', karma=-1)
+        upd.disable_autopush_for_critical_update()
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertTrue(up.has_negative_karma)
+        self.assertEquals(up.karma, -1)
+        self.assertEquals(up.status, UpdateStatus.testing)
+        self.assertEquals(up.request, None)
+        self.assertEquals(up.autokarma, False)
+
```

Comment:
You also need tests on ```check_karma_thresholds``` since the critical logic for this is in that method. Make sure to add both negative and positive tests so it is well covered.